src/syft/grid/webrtc_duet.py

```diff
@@ -0,0 +1,313 @@
+            # This will trigger the pull async task to be check signaling notifications
+            self._pull_msg_queue.put_nowait(
+                OfferPullRequestMessage(
+                    address=self.signaling_client.address,
+                    target_peer=address,
+                    host_peer=self.node.address,
+                    reply_to=self.signaling_client.address,
+                )
+            )
+        else:
+            # Push a WebRTC offer request to the address.
+            self.send_offer(address=address)
+
+        # This flag is used in order to finish the signaling process gracefully
+        # While self._available is True, the pull/push tasks will be running
+        # This flag will be setted to false when:
+        # 1 - End of the signaling process (checked by _update_availability()).
+        # 2 - Any Exception raised during these tasks.
+        self._available = True
+
+        # This attribute will be setted during the signaling messages exchange,
+        # and used to create a SoloRoute for the both sides.
+        self._client_metadata = ""
+
+        # Start async tasks and wait until one of them finishes.
+        # As mentioned before, these tasks can be finished by two reasons:
+        # 1 - Signaling process ends
+        # 2 - Unexpected Exception
+        asyncio.run(self.notify())
+
+        # If client_metadata != None, then the connection was created successfully.
+        if self._client_metadata:
+            # Deserialize client's metadata in order to obtain
+            # PySyft's location structure
+            (
+                spec_location,
+                name,
+                client_id,
+            ) = DomainClient.deserialize_client_metadata_from_node(
+                metadata=self._client_metadata
+            )
+
+            # Create a SoloRoute
+            route = SoloRoute(destination=spec_location, connection=self.connection)
+
+            # Intialize the super class
+            super().__init__(
+                domain=spec_location,
+                name=name,
+                routes=[route],
+                signing_key=self.signing_key,
+                verify_key=self.verify_key,
+            )
+            self.connection._client_address = self.address  # type: ignore
+        # If client_metada is None, then an exception occurred during the process
+        # The exception has been caught and saved in self._exception
+        else:
+            # NOTE: Maybe we should create a custom exception type.
+            raise Exception(
+                f"Something went wrong during the Duet init process. {self._exception}"
+            )
+
+    @syft_decorator(typechecking=True)
+    async def notify(self) -> None:
+        # Enqueue Pull/Push async tasks
+        push_task = asyncio.ensure_future(self.push())
+        pull_task = asyncio.ensure_future(self.pull())
+
+        # Wait until one of them finishes
+        done, pending = await asyncio.wait(
```

Comment:
What happens when we have a push and a pull that both need to be executed at the same time?

Reply:
Actually, they'll be executing concurrently from the beginning (We did it on the lines above).
The idea here is just to finish them gracefully if another one finishes (which can be triggered an exception or just the end of the signaling process.) :)